### ISSUE TITLE
Losing data when passing data through pipes

### ISSUE THREAD

**Issue by judofyr** [socketry/async-io#23]
```ruby
# wat.rb
require 'async'
require 'async/io'
require 'digest/sha1'
require 'securerandom'

Async.run do |task|
  r, w = IO.pipe.map { |io| Async::IO.try_convert(io) }

  task.async do |subtask|
    s = Digest::SHA1.new
    l = 0
    100.times do
      bytes = SecureRandom.bytes(4000)
      s << bytes
      w << bytes
      l += bytes.bytesize
    end
    w.close
    p [:write, l, s.hexdigest]
  end

  task.async do |subtask|
    s = Digest::SHA1.new
    l = 0
    while b = r.read(4096)
      s << b
      l += b.bytesize
    end
    p [:read, l, s.hexdigest]
  end
end
```

Running this (on macOS 10.14.3) gives me:

```
[:write, 400000, "73e44f9cfcdc7f50efe66aad2bff453b647bc722"]
[:read, 387680, "533edf52163ed91813b548aeb8ce23428029e4c0"]
```

This is a reduced test case. What I'm actually trying to do is to read a file from the internet (using async-http) and writing it to a pipe that's being read by archive-zip in order to stream a zip file.

**Comment by ioquatix** [comment]
Thanks, I'll investigate.

**Comment by ioquatix** [comment]
If you change the line in the first task:

```
puts (w << bytes)
```

You will see some times, less than 4000 bytes is written. This is expected, the pipe became full.

Do you think `<<` should guarantee all data is written?

I feel like that's a reasonable assumption, but right now it's an alias for `write` I guess, which can only write as much as there is space left in the pipe buffer.

**Comment by judofyr** [comment]
Hm. Maybe I'm misunderstanding how pipes work, but I was hoping that the write would become blocking once the pipe-buffer is full, causing the task to be paused, and then it would proceed once the reader caught up.

**Comment by judofyr** [comment]
I guess I'm confused that Stream#write doesn't follow the behavior of IO#write, but rather IO#write_nonblock.

**Comment by ioquatix** [comment]
`IO#write` might not write the full amount.

If you want buffering, you can use `Async::IO::Stream`.

**Comment by judofyr** [comment]
Oh. I thought `IO#write` guaranteed that the whole data was sent (blocking until the IO becomes writable again). I don't really want buffering, I just want it to block until the pipe/IO becomes writable again. Looks like `Async::IO::Stream` is going to solve my problem at the moment though.

**Comment by ioquatix** [comment]
https://ruby-doc.org/core-2.0.0/IO.html#method-c-write

> returns the length written.

https://ruby-doc.org/core-2.0.0/IO.html#method-i-3C-3C

It seems to internally buffer and write the entire string.

**Comment by judofyr** [comment]
I'm pretty sure people use `IO#write` with the intention that it writes the whole string (and blocks until it's completely flushed). Isn't that behavior the whole reason we have `IO#write_nonblock`?

**Comment by ioquatix** [comment]
I don't disagree with you, but it's simply how the lower level API works. Whether that's something a user wants to deal with is a good question though. The difference between `write` and `write_nonblock` is just that `write_nonblock` returns immediately if data couldn't be read/written. But it still has the same semantic regarding written data and returning the amount written.

**Comment by judofyr** [comment]
Are you sure? I've just never seen anyone checking the return value of `write/<<` and retrying/failing. E.g: https://github.com/rubyzip/rubyzip/blob/249775f5637e6d65112574b3ac1763dc2393c7f6/lib/zip/central_directory.rb#L51. Or is there a difference between `IO#write` and `IO#<<`?

**Comment by ioquatix** [comment]
`IO#<<` is different. But in `Async::IO` it is an alias for `write` which I believe is wrong. I've been playing around with fixing it. `<<` does a buffered write and might invoke the `write` syscall several times.

**Comment by judofyr** [comment]
Oh. That makes sense. I still can figure out how to trigger a case where `IO#write` does a partial write though. This code just hangs (because `#write` ends up blocking):

```ruby
require 'securerandom'

r, w = IO.pipe

1000.times do
  data = SecureRandom.bytes(100)
  if w.write(data) != 100
    raise "failed writing"
  end
end

p :done
```

**Comment by ioquatix** [comment]
Here is what I found.

https://stackoverflow.com/questions/12795842/under-what-circumstances-may-cs-write-function-write-less-data-than-requested

**Comment by judofyr** [comment]
Interesting. That means that async-io is not currently following the behavior of MRI though. I understand that async-io might have its own behavior, but it would be neat if I could use any regular blocking Ruby library (like rubyzip or archive-zip) and give it an `Async::IO`-object and everything would work (but now only blocking the fiber).

**Comment by ioquatix** [comment]
Yeah I think this is a bug.

**Comment by ioquatix** [comment]
Okay, so I played around with it today.

Firstly, I agree that the current situation is confusing.

I looked up the man pages and while `write` does seem to have some reasonably well documented behaviour, `send` is more ambiguous and I think that's where some of my confusion comes from.

- http://man7.org/linux/man-pages/man2/send.2.html
- http://man7.org/linux/man-pages/man2/write.2.html

> The number of bytes written may be less than count if, for example,
       there is insufficient space on the underlying physical medium, or the
       RLIMIT_FSIZE resource limit is encountered (see setrlimit(2)), or the
       call was interrupted by a signal handler after having written less
       than count bytes.  (See also pipe(7).)

Do you think that is an exhaustive list? Or do you think there are some situations where `write` would write less than the specified amount of data?

The next question is regarding the design of this library.

Right now, `Async::IO` doesn't implement the limited semantics above (limited as in the several cases given where the returned size was less than the specified data size).

However, you can achieve this very easily using the buffering `Async::IO::Stream`.

The question is, should we do this by default?

I'm on the fence. But I certainly see both sides of the argument. For example `Async::IO::Generic#<<` has no way to indicate bytes written, so I think it's reasonable to assume it should try to write all data unless encountering interrupt or other exceptional situation.

But `Async::IO::Generic#write` is less clear to me. If the list in the man page is exhaustive, then I think we could reasonably assume the default implementation should try to write all data... but if the man page is not exhaustive - i.e. there are some situations where the OS returns less than requested bytes written, and that's normal/expected, then I don't believe we should change that.

**Comment by ioquatix** [comment]
Probably the most logical place for such a situation to show up, is when `write` is internally invoking `sendto`.

**Comment by ioquatix** [comment]
I was interested to see what other OS does:

https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.3.0/com.ibm.zos.v2r3.bpxbd00/rtwri.htm

Also POSIX: http://pubs.opengroup.org/onlinepubs/9699919799/functions/write.html

**Comment by judofyr** [comment]
I'm mainly interested in keeping the behavior equivalent to Ruby's built-in `IO#write` and `IO#<<`.

**Comment by ioquatix** [comment]
Ruby `io.c` implements it's own loop to handle this:

https://github.com/ruby/ruby/blob/trunk/io.c#L1411-L1415

I guess it is important to consider Ruby's semantics, but that's more complex given that it must be similar across MRI, TruffleRuby, JRuby, etc. So, I'm trying to take all that into account. If it's okay, by the POSIX specifications, to write less than the amount specified, then I'll probably stick with that just for efficiency purposes, but I totally see where you are coming from w.r.t. how Ruby behaves.

**Comment by ioquatix** [comment]
So, I think this issue equally applies to `IO#read` too. Here are my thoughts:

The logic of `blocking` IO is that if you want to read N bytes or write N bytes, the IO should block until that operation has completed or the underlying IO experiences a failure. I think that's a reasonable assumption to make.

The logic of non-blocking IO is to invoke some operation and read whatever is available, or write whatever is possible, without waiting for additional data to satisfy the full request (and in Ruby there is `readpartial` which is a bit of a hack).

The confusion for me is that in some cases, blocking IO can return less than requested, e.g. if the disk is full or the file has less than the requested amount of data.

```
[26] pry(main)> File.write("tmp.txt", "Hello World")                                                                                                                                           
=> 11
[27] pry(main)> File.open("tmp.txt"){|f| f.read(32)}                                                                                                                                           
=> "Hello World"
```

So semantically, it's not totally clean.

Ideally, we are not implementing semantics in our own code, and just deferring to the underlying implementation. But we can't do this in the case of `read`/`write` because we are implementing them in terms of nonblocking `read_nonblock`/`write_nonblock`. Therefore, we try to match the semantics of `read`/`write` and this is where the issue has happened.

Here are two implementations which try to bridge that gap:

```ruby
			# Read the specified number of bytes from the input stream.
			def read(length)
				# Fast path:
				buffer = self.read_nonblock(length)
				
				while buffer.bytesize < length
					if chunk = self.read_nonblock(length - buffer.bytesize)
						buffer << chunk
					else
						break
					end
				end
				
				return buffer
			end
			
			# Write entire buffer to output stream.
			def write(buffer)
				remaining = buffer.bytesize
				
				# Fast path:
				written = self.write_nonblock(buffer)
				return if written == remaining
				
				# Slow path:
				remaining -= written
				
				while remaining > 0
					wrote = self.write_nonblock(buffer.byteslice(written, remaining))
					
					remaining -= wrote
					written += wrote
				end
				
				return written
			end
```

WDYT?

**Comment by ioquatix** [comment]
One more question I wonder is how does `TCPSocket#read` behave?

**Comment by ioquatix** [comment]
Okay, so I'm convinced that you are correct about Ruby's behaviour.

The problem now is making it work without breaking existing assumptions about the methods.

**Comment by ioquatix** [comment]
I found the first issue:

`Async::IO::Generic#read` can't accept no arguments. But it should.

**Comment by ioquatix** [comment]
Okay, so I made a PR but it probably requires more testing.

**Comment by ioquatix** [comment]
Okay, I think I'm convinced, I'll merge this into master and see how many other things blow up (hopefully none).

**Comment by ioquatix** [comment]
Released in `v1.23.0`.

**Comment by judofyr** [comment]
Very nice!
